### PR TITLE
terrror: a tiny refinement on ErrorClass.String

### PR DESCRIPTION
**Pull request by spongedu**
Minimal ErrorClass lookup time by eliminating comparation in "switch .... case....".

### CODE REVIEW

File: terror/terror.go [hunk 1]
```diff
@@ -85,49 +85,34 @@ const (
...
-		return "table"
-	case ClassTypes:
-		return "types"
-	case ClassGlobal:
-		return "global"
-	case ClassMockTikv:
-		return "mocktikv"
+	s, exists := tm[ec]
+	if exists {
+		return s
```
Comment (zz-jason) [inline terror/terror.go]:
```go
if s, exists := tm[ec]; exists {
	return s
}
```

File: terror/terror.go [hunk 2]
```diff
@@ -85,49 +85,34 @@ const (
 	// Add more as needed.
 )
 
+var tm = map[ErrClass]string{
```
Comment (zz-jason) [inline terror/terror.go]: change `tm` to a more readable one

**Comment by zz-jason** [comment]
@hanfei1991 @XuHuaiyu @shenli PTAL

**Comment by tiancaiamao** [comment]
ErrClass is int type, so the switch case should be rather efficient.
Any benchmark can prove this change work? @spongedu

**Comment by tiancaiamao** [comment]
Benchmark shows, this change slower than before:

```
func BenchmarkString(b *testing.B) {
	ec := ClassMeta
	// ec := ClassOptimizer
	// ec := ClassMockTikv
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		_ = ec.String()
	}
}
```

```
BenchmarkString-4   	1000000000	         2.59 ns/op
```

after:

```
BenchmarkString-4   	100000000	        13.4 ns/op
```

Please be more careful when reviewing PR. @zz-jason @XuHuaiyu 
As a general rule, but fix should come with tests, performance tuning should have benchmark result, and watch for even a small typo...

**Comment by zz-jason** [comment]
@tiancaiamao  how about prepare a slice with random`ErrClass`s, and test for that slice ?

**Comment by tiancaiamao** [comment]
Actually, I don't mind the performance of this function, it shall not be called frequently on normal cases, so just keep it as it is.
Here I mean we should always keep in mind the cost for each kind of operations. @zz-jason

**Comment by spongedu** [comment]
@tiancaiamao  @zz-jason @XuHuaiyu  yep. I've done experiments myself and got similar results as yours. Table lookup seems more expensive than severial  branch prediction misses. I'll be more careful on this next time. Thx for you advice.

**Pull request merged**